### ISSUE TITLE
Move the old Pseudonamespaces to real namespaces

### ISSUE THREAD

**Issue by rengaw83** [DMKEBUSINESSGMBH/typo3-mksearch#51]
Rector does a lot of automated refactoring stuff.  
There are already a [PseudoNamespace To NamespaceRector](https://github.com/rectorphp/rector/blob/0b0cf24098cd779453966abb0e402e2007d16a8d/docs/AllRectorsOverview.md#pseudonamespacetonamespacerector) we can try to use.

**Comment by Baachi** [comment]
I tried to refactor this with rector, but i run into some problems. If run this rector the namespace would be as followed `tx\mksearch\engine\interface` and not `DMK\MKsearch\Engine\Interface`.

So i tried to run the [`RenameNamespaceRector`](https://github.com/rectorphp/rector/blob/0b0cf24098cd779453966abb0e402e2007d16a8d/docs/AllRectorsOverview.md#renamenamespacerector) which doesn't have any change. 

I try to create PR next week, so we have a good starting point.

**Comment by fchaussin** [comment]
Hi @Baachi , did you managed to refactor the Classes with Namspaces? I can help...

**Comment by Baachi** [comment]
Hey @fchaussin,

First of all, thanks for your interest and help! :+1: 

I had two main problems regarding this issue. First of all the usage of the `PseudoNamespaceToNamespaceRector` and `RenameNamespaceRector` had some bugs that the resulting namespace wasn't that what I expected. As far as I remember the result was something like this `tx\mksearch\engine\interface`, which isn't valid in PHP, because `interface` is a keyword.

The second problem which is more complex is, that the current structure doesn't match with the current state of the art. So rename the namespace isn't enough. For example, [this](https://github.com/DMKEBUSINESSGMBH/typo3-mksearch/tree/master/service/engine) folder/namespace should be converted to `DMK\MKSearch\Indexer` and the [corresponding interface](https://github.com/DMKEBUSINESSGMBH/typo3-mksearch/blob/master/interface/class.tx_mksearch_interface_SearchEngine.php) should be moved to this namespace/folder, too.

And last but not least, we need to provide a class map for TYPO3, so we don't introduce a BC break. 

All in all, not a big issue, but requires some time and effort.

**Comment by fchaussin** [comment]
> Hey @fchaussin,
> 
> First of all, thanks for your interest and help! 👍
> 
> I had two main problems regarding this issue. First of all the usage of the `PseudoNamespaceToNamespaceRector` and `RenameNamespaceRector` had some bugs that the resulting namespace wasn't that what I expected. As far as I remember the result was something like this `tx\mksearch\engine\interface`, which isn't valid in PHP, because `interface` is a keyword.
> 
> The second problem which is more complex is, that the current structure doesn't match with the current state of the art. So rename the namespace isn't enough. For example, [this](https://github.com/DMKEBUSINESSGMBH/typo3-mksearch/tree/master/service/engine) folder/namespace should be converted to `DMK\MKSearch\Indexer` and the [corresponding interface](https://github.com/DMKEBUSINESSGMBH/typo3-mksearch/blob/master/interface/class.tx_mksearch_interface_SearchEngine.php) should be moved to this namespace/folder, too.
> 
> And last but not least, we need to provide a class map for TYPO3, so we don't introduce a BC break.
> 
> All in all, not a big issue, but requires some time and effort.

Yes that is what I encourted yesterday afternoon, it is a pretty long refactoring, and I will be only able to test the Elastic side. By the way, refactoring namespace is probably the starting point of following the new TYPO3 Extension conventions, and leaving RN_Base dependency. And maybe using PSR-14 based EventDispatcher instead of Hooks...

**Comment by Baachi** [comment]
@fchaussin There aren't any plans for v10 but I agree with you, that we can create a better library with all the new shiny features (EventDispatcher, Middleware, DependencyInjection) that come with the new TYPO3 release. 

But the main selling point for this library (at least in our opinion) is the compatibility with at least the last two LTS releases (currently it is compatible with v6 - v9). And introduce these features would mean, that we create a completely new library, which requires a bunch of work and effort.

**Comment by fchaussin** [comment]
> @fchaussin There aren't any plans for v10 but I agree with you, that we can create a better library with all the new shiny features (EventDispatcher, Middleware, DependencyInjection) that come with the new TYPO3 release.
> 
> But the main selling point for this library (at least in our opinion) is the compatibility with at least the last two LTS releases (currently it is compatible with v6 - v9). And introduce these features would mean, that we create a completely new library, which requires a bunch of work and effort.

I totally agree with you, 1st step is to keep it work even with deprececiations. The refactoring is too large to guarantee backward compatibility without too much headache! Reusing the code in a brand new project with the new TYPO3 conventions should be a little easier.

**Comment by neufeind** [comment]
It would be really great to get this going. Since there are several things to redo besides namespaces imho a separate major-version without the aim to provide backward-compatibility (classmap, ...) might be our best bet. Other extensions had similar factorings where a migration-help (what changed, from what to what etc.) would help. I think that's viable for integrators moving to a new major version.

**Comment by hannesbochmann** [comment]
In the moment there are no plans to do further development of mksearch. So a refactoring to use namespaces won't happen right now. There will be no new features too, as long as nobody sponsors the development. But we will keep mksearch alive so there will be compatible versions for future TYPO3 versions.
